Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -65,6 +69,122 @@
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty()) {
+        completionHandler(nullptr);
+        return;
+    }
+
+    String matchingUniqueIdentifier;
+    String storageDirectory;
+    String displayName;
+
+    auto recordHolder = RecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        completionHandler(recordHolder->recordsMap.takeFirst());
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+        auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);
+
+        URL lastBaseURL;
+        if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath, displayName, lastBaseURL))
+            continue;
+
+        if (this->extensionContext(lastBaseURL)->extension() == extensionContext.extension()) {
+            matchingUniqueIdentifier = uniqueIdentifier;
+            break;
+        }
+    }
+
+    if (!matchingUniqueIdentifier) {
+        completionHandler(nullptr);
+        return;
+    }
+
+    for (auto type : types) {
+        calculateStorageSizeForDataType(type, storageDirectory, matchingUniqueIdentifier, makeBlockPtr([recordHolder, aggregator, matchingUniqueIdentifier, displayName, type](size_t size) mutable {
+            Ref record = recordHolder->recordsMap.ensure(matchingUniqueIdentifier, [&] {
+                return WebExtensionDataRecord::create(displayName, matchingUniqueIdentifier);
+            }).iterator->value;
+            record->setSizeOfType(type, size);
+        }));
+    }
+}
+
+void WebExtensionController::removeData(OptionSet<WebExtensionDataType> types, const Vector<Ref<WebExtensionDataRecord>>& records, CompletionHandler<void()>&& completionHandler)
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty() || records.isEmpty()) {
+        completionHandler();
+        return;
+    }
+
+    auto aggregator = MainRunLoopCallbackAggregator::create([completionHandler = WTFMove(completionHandler)]() mutable {
+        completionHandler();
+    });
+
+    for (auto& record : records) {
+        auto uniqueIdentifier = record.get().uniqueIdentifier();
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+
+        for (auto dataType : record.get().types()) {
+            // FIXME: <https://webkit.org/b/269100> Remove window.localStorage, window.sessionStorage, indexedDB.
+            auto *storage = [[_WKWebExtensionStorageSQLiteStore alloc] initWithUniqueIdentifier:uniqueIdentifier storageType:dataType directory:storageDirectory usesInMemoryDatabase:NO];
```

Comment:
This could be a helper method that returns `_WKWebExtensionStorageSQLiteStore *`, which also can be used below in `calculateStorageSizeForDataType()`.

This should also continue early, or have the helper return `nil` if `dataType` is not `Local` or `Sync`, until the FIXME is fixed. Since `usesInMemoryDatabase:NO` will be wrong for `Session`.

Reply:
yea passing `usesInMemoryDatabase:NO` is wrong here. fixed this to pass in `dataType::WebExtensionDataTypeSession` for `usesInMemoryDatabase`

Reply:
@kiaraarose But we will need to use the right session DB, from the loaded `WebExtensionContext`, since a new one will just be empty. So it is likely best to not do anything when type is Session for now.

Reply:
True. But since storage used for Session will be cleared when a context unloads, I'll just add a check to see if the context is loaded and then use the `sessionStorageStore()` on the WebExtensionContext to handle calculating/removing storage